network/p2p/p2pbuilder/gossipsub/gossipSubBuilder.go

```diff
@@ -0,0 +1,231 @@
+}
+
+func (g *Builder) SetAppSpecificScoreParams(f func(peer.ID) float64) {
+	g.peerScoringParameterOptions = append(g.peerScoringParameterOptions, scoring.WithAppSpecificScoreFunction(f))
+}
+
+func NewGossipSubBuilder(logger zerolog.Logger, metrics module.GossipSubMetrics) *Builder {
+	return &Builder{
+		logger:                      logger.With().Str("component", "gossipsub").Logger(),
+		metrics:                     metrics,
+		gossipSubFactory:            defaultGossipSubFactory(),
+		gossipSubConfigFunc:         defaultGossipSubAdapterConfig(),
+		peerScoringParameterOptions: make([]scoring.PeerScoreParamsOption, 0),
+	}
+}
+
+func defaultGossipSubFactory() p2p.GossipSubFactoryFunc {
+	return func(ctx context.Context, logger zerolog.Logger, h host.Host, cfg p2p.PubSubAdapterConfig) (p2p.PubSubAdapter, error) {
+		return p2pnode.NewGossipSubAdapter(ctx, logger, h, cfg)
+	}
+}
+
+func defaultGossipSubAdapterConfig() p2p.GossipSubAdapterConfigFunc {
+	return func(cfg *p2p.BasePubSubAdapterConfig) p2p.PubSubAdapterConfig {
+		return p2pnode.NewGossipSubAdapterConfig(cfg)
+	}
+}
+
+// Build creates a new GossipSub pubsub system.
+// It returns the newly created GossipSub pubsub system and any errors encountered during its creation.
+// Arguments:
+// - ctx: the irrecoverable context of the node.
+//
+// Returns:
+// - p2p.PubSubAdapter: a GossipSub pubsub system for the libp2p node.
+// - p2p.PeerScoreTracer: a peer score tracer for the GossipSub pubsub system (if enabled, otherwise nil).
+// - error: if an error occurs during the creation of the GossipSub pubsub system, it is returned. Otherwise, nil is returned.
+// Note that on happy path, the returned error is nil. Any error returned is unexpected and should be handled as irrecoverable.
+func (g *Builder) Build(ctx irrecoverable.SignalerContext) (p2p.PubSubAdapter, p2p.PeerScoreTracer, error) {
+	gossipSubConfigs := g.gossipSubConfigFunc(&p2p.BasePubSubAdapterConfig{
+		MaxMessageSize: p2pnode.DefaultMaxPubSubMsgSize,
+	})
+	gossipSubConfigs.WithMessageIdFunction(utils.MessageID)
+
+	if g.rsys == nil {
+		return nil, nil, fmt.Errorf("could not create gossipsub: routing system is nil")
+	}
+	gossipSubConfigs.WithRoutingDiscovery(g.rsys)
+
+	if g.subscriptionFilter != nil {
+		gossipSubConfigs.WithSubscriptionFilter(g.subscriptionFilter)
+	}
+
+	var scoreOpt *scoring.ScoreOption
+	var scoreTracer p2p.PeerScoreTracer
+	if g.gossipSubPeerScoring {
+		scoreOpt = scoring.NewScoreOption(g.logger, g.idProvider, g.peerScoringParameterOptions...)
+		gossipSubConfigs.WithScoreOption(scoreOpt)
+
+		if g.gossipSubScoreTracerInterval > 0 {
+			scoreTracer = tracer.NewGossipSubScoreTracer(
+				g.logger,
+				g.idProvider,
+				g.metrics,
+				g.gossipSubScoreTracerInterval)
+			gossipSubConfigs.WithScoreTracer(scoreTracer)
+			g.logger.Debug().Msg("starting gossipsub score tracer")
+			scoreTracer.Start(ctx)
+			<-scoreTracer.Ready()
```

Comment:
This seems like it should be done by the caller. At the very least, it makes it more obvious that we may want to wait on `scoreTracer.Done()` before returning from the component worker.

Reply:
To address your comment, I made the `PubSubAdapter` a startable component and move all gossipsub startup logic there: https://github.com/onflow/flow-go/pull/3986/commits/0600d5c82f5a8ddeb311911a316826b08b0ac514